### PR TITLE
Opacity

### PR DESCRIPTION
**Pull request by Vedhus**
Pull request for code review

### CODE REVIEW

**Comment by Kailang** [comment]
Checklist:
- Inconsistent coding style.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawSelect.cpp#L105

1: `){` should be `) {`

File: Qt/InstaDam/Selector/label.cpp [hunk 1]
```diff
@@ -148,23 +148,27 @@ void Label::clear() {
...
-         polygonItems != polygonObjects.end(); ++polygonItems)
+
+    for (polygonItems = polygonObjects.begin(); polygonItems != polygonObjects.end(); ++polygonItems)
         polygonItems.value()->setOpacity(normalizedValue);
-    qInfo("Opacity Polygon\n");
+
+    for (ellipseItems = ellipseObjects.begin(); ellipseItems != ellipseObjects.end(); ++ellipseItems)
+        ellipseItems.value()->setOpacity(normalizedValue);
+
+    for (freeDrawItems = freeDrawObjects.begin(); freeDrawItems != freeDrawObjects.end(); ++freeDrawItems)
```
Comment (Kailang) [inline Qt/InstaDam/Selector/label.cpp]: 3: The same for loop is duplicated 4 times.

File: Qt/InstaDam/Test/app/testLabel.cpp [hunk 2]
```diff
@@ -69,6 +69,9 @@ void TestSelect::testLabelSetOpacity() {
     generateData();
     myLabel->setOpacity(2);
     QCOMPARE(ritem->SelectItem::opacity(), .02);
+    QCOMPARE(eitem->SelectItem::opacity(), .02);
```
Comment (Kailang) [inline Qt/InstaDam/Test/app/testLabel.cpp]: 2: Test incomplete.
This test does not cover the changes in FreeDrawErase class.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawErase.h#L28

4: `paint()` in FreeDrawErase refuses the implementation in its parent class.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L36
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L38

1: The indention is not aligned.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L166

3: The same for loop is duplicated 4 times.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L208

3: The same if-for structure is duplicated 3 times.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L273

3: The same if-while structure is duplicated 4 times.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L341
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L394

3: The similar structure is duplicated 4 times.

File: Qt/InstaDam/Selector/ellipseSelect.h [hunk 3]
```diff
@@ -35,6 +35,7 @@ class EllipseSelect : public QGraphicsEllipseItem, public BoxBasedSelector {
     void setRectUnchecked(QRectF rect) override;
     void updatePen(QPen pen) override;
     void toPixmap(QPainter *painter) override;
+    void setOpacity(qreal val);
```
Comment (Kailang) [inline Qt/InstaDam/Selector/ellipseSelect.h]: 5: Cannot find the User Story that the change implements.

File: Qt/InstaDam/Selector/freeDrawErase.h [hunk 4]
```diff
@@ -32,7 +32,7 @@ class FreeDrawErase : public FreeDrawSelect {
         UNUSED(widget);
     }
     void toPixmap(QPainter *painter) override {UNUSED(painter);}
-
+    void setOpacity(qreal val) {UNUSED(val);}
```
Comment (Kailang) [inline Qt/InstaDam/Selector/freeDrawErase.h]: 5: Cannot find the User Story that the change implements.

File: Qt/InstaDam/Selector/freeDrawSelect.h [hunk 5]
```diff
@@ -34,6 +34,7 @@ class FreeDrawSelect : public QGraphicsPixmapItem, public SelectItem {
         return FreeDrawSelect::baseInstruction;
     }
     QRectF boundingRect() const override;
+    void setOpacity(qreal val);
```
Comment (Kailang) [inline Qt/InstaDam/Selector/freeDrawSelect.h]: 5: Cannot find the User Story that the change implements.

File: Qt/InstaDam/Selector/selectItem.h [hunk 6]
```diff
@@ -80,6 +80,7 @@ class SelectItem : public QGraphicsItem {
     virtual void resizeItem(const int vertex, QPointF &shift) = 0;
     virtual void rotate(const QPointF &from, const QPointF &to) = 0;
     virtual void updatePen(QPen pen) = 0;
+    void setOpacity(qreal val) {QGraphicsItem::setOpacity(val);}
```
Comment (Kailang) [inline Qt/InstaDam/Selector/selectItem.h]: 5: Cannot find the User Story that the change implements.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawSelect.cpp#L209
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawSelect.cpp#L240

6: Magic string `objectID`.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L35

1: Inconsistent indention with other places.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/selectItem.h#L64

6: Magic number `10`.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/ellipseSelect.h#L11

8: Class is not in any namespace.

**Comment by Kailang** [comment]
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/polygonSelect.h#L9
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawErase.h#L14
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawSelect.h#L12
https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/selectItem.h#L38

8: Classes are not in any namespace.

**Comment by Vedhus** [comment]
> [InstaDam/Qt/InstaDam/Selector/label.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L36)
> 
> Line 36 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  readServer(json); 
> 
> [InstaDam/Qt/InstaDam/Selector/label.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L38)
> 
> Line 38 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  read(json); 
> 1: The indention is not aligned.



> [InstaDam/Qt/InstaDam/Selector/freeDrawSelect.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawSelect.cpp#L105)
> 
> Line 105 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  : QGraphicsPixmapItem(item), SelectItem(label, item){ 
> 1: `){` should be `) {`

Thanks, will add the space.

**Comment by Vedhus** [comment]
> [InstaDam/Qt/InstaDam/Selector/label.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L36)
> 
> Line 36 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  readServer(json); 
> 
> [InstaDam/Qt/InstaDam/Selector/label.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L38)
> 
> Line 38 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  read(json); 
> 1: The indention is not aligned.

Will fix indentation.

File: Qt/InstaDam/Selector/label.cpp (hunk 1 above)
Comment (astro-friedel) [reply Qt/InstaDam/Selector/label.cpp]: There is a for loop for each if the distinct Hashes, there is no way to shorten them as each contains a different data type.

**Comment by astro-friedel** [comment]
> [InstaDam/Qt/InstaDam/Selector/label.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L273)
> 
> Line 273 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  if (!freeDrawObjects.isEmpty()) { 
> 3: The same if-while structure is duplicated 4 times.

This code is looping through the different data hashes, while they may call similarly named methods, they hashes are distinct and cannot be combined.

**Comment by astro-friedel** [comment]
> [InstaDam/Qt/InstaDam/Selector/label.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L166)
> 
> Line 166 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  for (freeDrawItems = freeDrawObjects.begin(); freeDrawItems != freeDrawObjects.end(); ++freeDrawItems) 
> 3: The same for loop is duplicated 4 times.

There is a for loop for each if the distinct Hashes, there is no way to shorten them as each contains a different data type.

**Comment by astro-friedel** [comment]
> [InstaDam/Qt/InstaDam/Selector/label.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L208)
> 
> Line 208 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  for (QJsonArray::iterator it = polyArray.begin(); it != polyArray.end(); 
> 3: The same if-for structure is duplicated 3 times.

This code is reading in different Json objects, one for each data type. They are each separated by if statements looking for the existence of the specific data structures before looping over the contents. They have to be in separate loops.

File: Qt/InstaDam/Test/app/testLabel.cpp (hunk 2 above)
Comment (astro-friedel) [reply Qt/InstaDam/Test/app/testLabel.cpp]: The FreeDrawErase class does not have an opacity as it is never painted. The class is designed to hold brush strokes that have been removed from FreeDrawSelect classes for undo and redo purposes only. The data from FreeDrawErase instances are not saved beyond the exiting of the program.

**Comment by astro-friedel** [comment]
> [InstaDam/Qt/InstaDam/Selector/freeDrawErase.h](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawErase.h#L28)
> 
> Line 28 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  void paint(QPainter *painter, const QStyleOptionGraphicsItem *option, 
> 4: `paint()` in FreeDrawErase refuses the implementation in its parent class.

The FreeDrawErase class is never painted. The class is designed to hold brush strokes that have been removed from FreeDrawSelect classes for undo and redo purposes only. The data from FreeDrawErase instances are not saved beyond the exiting of the program.

**Comment by astro-friedel** [comment]
> [InstaDam/Qt/InstaDam/Selector/freeDrawSelect.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawSelect.cpp#L209)
> 
> Line 209 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  myID = json["objectID"].toInt(); 
> 
> [InstaDam/Qt/InstaDam/Selector/freeDrawSelect.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/freeDrawSelect.cpp#L240)
> 
> Line 240 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  json["objectID"] = myID; 
> 6: Magic string `objectID`.

We will implement this, it was on the to-do list

**Comment by astro-friedel** [comment]
> [InstaDam/Qt/InstaDam/Selector/ellipseSelect.h](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/ellipseSelect.h#L11)
> 
> Line 11 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  class EllipseSelect : public QGraphicsEllipseItem, public BoxBasedSelector { 
> 8: Class is not in any namespace.

This project is small enough that namespaces for these objects are not needed.

File: Qt/InstaDam/Selector/freeDrawSelect.h (hunk 5 above)
Comment (Vedhus) [reply Qt/InstaDam/Selector/freeDrawSelect.h]: Please see the revised user story list here (User Story 37)

File: Qt/InstaDam/Selector/selectItem.h (hunk 6 above)
Comment (Vedhus) [reply Qt/InstaDam/Selector/selectItem.h]: Please see the revised user story list here (User Story 37) https://wiki.illinois.edu/wiki/display/CS428SP19/IDM+-+Iteration+2

File: Qt/InstaDam/Selector/freeDrawErase.h (hunk 4 above)
Comment (Vedhus) [reply Qt/InstaDam/Selector/freeDrawErase.h]: Please see the revised user story list here (User Story 37) https://wiki.illinois.edu/wiki/display/CS428SP19/IDM+-+Iteration+2

**Comment by Vedhus** [comment]
> [InstaDam/Qt/InstaDam/Selector/label.cpp](https://github.com/Vedhus/InstaDam/blob/94de7f494efedf6183c819349c366d3052fc9e7e/Qt/InstaDam/Selector/label.cpp#L35)
> 
> Line 35 in [94de7f4](/Vedhus/InstaDam/commit/94de7f494efedf6183c819349c366d3052fc9e7e)
> 
>  if(server) 
> 1: Inconsistent indention with other places.

Will change the indentation

**Comment by astro-friedel** [comment]
> Checklist (things to avoid):
> 
> 1. Inconsistent coding style.
> 2. Tests do not test the changed code.
> 3. Duplicated code.
> 4. Refused bequest.
> 5. Changes do not fulfill the requirements.
> 6. Use magic constants.
> 7. Lines are too long.
> 8. Does not use namespaces.
> 9. Does not use member initializer list when possible.
> 10. Use `0` or `NULL` instead of `nullptr`.

Using 0 and NULL are deprecated, 0 can be recast as an int under some circumstances and NULL is just a value not a type. They have been replaced by nullptr, which is a true pointer.

**Comment by astro-friedel** [comment]
Indentations and other formatting issues have been fixed.
'Magic' items have been converted to constants.
Since this was merged to master, I have pushed these changes to the HW4fixes branch.

**Pull request merged**